File: day3/day3.kt
import java.io.File

fun main() {
    val input = File("day3.txt").readLines()
    fun count(dx: Int, dy: Int) = (0 until input.size step dy).count {
        input[it][(it * dx / dy) % input[0].length] == '#' }
    println(count(3, 1))
    println(listOf(1 to 1, 3 to 1, 5 to 1, 7 to 1, 1 to 2).map {
        (dx, dy) -> count(dx, dy) }.fold(1L) { a, b -> a * b })
}

File: day3/day3.py
def count(dx, dy):
    res, x, y = 0, 0, 0
    while y < len(grid):
        res += grid[y][x] == '#'
        y += dy
        x = (x + dx) % len(grid[0])
    return res

grid = [x.rstrip() for x in open("day3.txt")]
print(count(3, 1))
print(count(1, 1) * count(3, 1) * count(5, 1) * count(7, 1) * count(1, 2))
